rust/azure_iot_operations_connector/src/base_connector/managed_azure_device_registry.rs

```diff
@@ -38,57 +43,342 @@ where
+            let asset_client_creation_observation = AssetClientCreationObservation {
+                asset_create_observation,
+                connector_context: self.connector_context.clone(),
+            };
+
+            // and then get device update observation as well and turn it into a DeviceEndpointClientUpdateObservation
+            let device_endpoint_client_update_observation =  match Retry::spawn(RETRY_STRATEGY, async || -> Result<DeviceUpdateObservation, RetryError<azure_device_registry::Error>> {
+                self.connector_context
+                    .azure_device_registry_client
+                    .observe_device_update_notifications(
+                        device_endpoint_ref.device_name.clone(),
+                        device_endpoint_ref.inbound_endpoint_name.clone(),
+                        self.connector_context.default_timeout,
+                    )
+                    // retry on network errors, otherwise don't retry on config/dev errors
+                    .await.map_err(observe_error_into_retry_error)
+            }).await {
+                Ok(device_update_observation) => {
+                    DeviceEndpointClientUpdateObservation {
+                        device_update_observation,
+                        connector_context: self.connector_context.clone(),
+                    }
+                },
+                Err(e) => {
+                  log::error!("Failed to observe for device update notifications after retries: {e}");
+                  log::error!("Dropping device endpoint create notification: {device_endpoint_ref:?}");
+                  continue;
+                },
+            };
+
+            // get the device definition
+            let device = match Retry::spawn(
+                RETRY_STRATEGY,
+                async || -> Result<Device, RetryError<azure_device_registry::Error>> {
+                    self.connector_context
+                        .azure_device_registry_client
+                        .get_device(
+                            device_endpoint_ref.device_name.clone(),
+                            device_endpoint_ref.inbound_endpoint_name.clone(),
+                            self.connector_context.default_timeout,
+                        )
+                        .await
+                        .map_err(|e| {
+                            match e.kind() {
+                                // network/retriable
+                                azure_device_registry::ErrorKind::AIOProtocolError(_) => {
+                                    RetryError::transient(e)
+                                }
+                                _ => {
+                                    // ServiceError indicates an error in the configuration, so we want to get a new notification instead of retrying this operation
+                                    // InvalidRequestArgument shouldn't be possible since timeout is already validated
```

Comment:
Might be worth formalizing this logic with an unreachable! macro

Reply:
good point, updated. Didn't move the duplicate observation error yet as since the implementation is still in progress, I'm not sure this is completely guaranteed yet (but will be moved to be unreachable in the future, and that's documented)